### DISCUSSION TITLE
What is the guideline to add/test a new cloud provider

### DISCUSSION THREAD

**Discussion by yuhuishi-convect** [nebari-dev/nebari#1090] [Q&A] [answered]
Hello, I have tested using qhub and really happy about its convenience to spawn jupyterhub on various cloud providers.
I would like to contribute to adding a few more cloud providers -- alicloud, tencent cloud more specifically. They all provide managed Kubernetes services so I assume the upper layer (applications) should look pretty much the same even running on different cloud providers. The infra layer (to configure and initialize Kubernetes clusters) shall be different.

What is the general guideline on adding and testing a new cloud provider? Is there some reference pull request I can quickly check to get some sense?
Thanks!

**Comment by costrouc** [comment, marked as answer]
Hey @yuhuishi-convect glad to hear that QHub has been working well for you! We don't have exact docs at the moment for doing this but you are right. In general all we need is a terraform remote state, a Kubernetes cluster with the ability to create a load-balancer. A lot has changed since this PR https://github.com/Quansight/qhub/pull/314 but this is where we added Azure. 

We are about to have a 0.4 release that should make this much easier to see. There are two stages:
 - https://github.com/Quansight/qhub/tree/main/qhub/template/stages/01-terraform-state
 - https://github.com/Quansight/qhub/tree/main/qhub/template/stages/02-infrastructure

These are the two stages that will be to be implemented. I will likely have missed a few small pieces but otherwise these are the main steps.
